minject/inject.py

```diff
@@ -103,6 +148,11 @@ def __init__(self, key: "RegistryKey[T_co]") -> None:
     def resolve(self, registry_impl: Resolver) -> T_co:
         return registry_impl.resolve(self._key)
 
+    async def aresolve(self, registry_impl: Resolver) -> T_co:
+        if _is_key_async(self._key):
+            return await registry_impl._aresolve(self._key)
+        return registry_impl.resolve(self._key)
```

Comment:
This assumes that the all `Resolver.resolve` implementations are strictly non-blocking, but there is currently no such constraint documented on that method.

Should we call this using `asyncio.to_thread` instead, or perhaps require the `_aresolve` method implementation itself to support async keys?

Reply:
Interesting problem, I'm leaning toward "don't run in `asyncio.to_thread`" but I think I could be swayed in either direction.

Here is my thinking:

I think `minject` strives to be as "unmagical" as possible. I think `asyncio.to_thread` could fix an issue the user added to their code "run blocking code in event loop". I think this would potentially make the user unaware of the issue at initialization (`minject` would fix/hide the problem), and allow the user to have incorrect async/sync interactions at runtime.

Do people think this makes sense, or does this seem a bit off?

Reply:
@xiyan128 also commented on this here: https://github.com/duolingo/minject/pull/49#discussion_r1784590345

Reply:
I've moved all synchronous `resolve` + `close` calls into `asyncio.to_thread` when called from `a*` functions/methods.

I'm slightly nervous that this could encourage people to run blocking code within their async classes (`minject` initialization will suggest that a async -> sync call causes no issues as `minject` is handling the `asyncio.to_thread` call on the users behalf), but overall I think this is correct as there as I'm not sure you'd be able to properly do async -> sync initialization without this feature.